keepalived-vip/controller.go

```diff
@@ -0,0 +1,159 @@
+package main
+
+import (
+	"fmt"
+	"os"
+	"reflect"
+	"sort"
+	"text/template"
+	"time"
+
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/kubernetes/pkg/util/exec"
+	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/kubernetes/pkg/util/wait"
+
+	"github.com/golang/glog"
+)
+
+type keepalivedController struct {
+	clientset  *kubernetes.Clientset
+	keepalived *keepalived
+
+	namespace   string
+	serviceName string
+	podName     string
+
+	template *template.Template
+	config   map[string]interface{}
+}
+
+func newKeepalivedController(clientset *kubernetes.Clientset, namespace, serviceName, podName string) (*keepalivedController, error) {
+	c := &keepalivedController{
+		clientset:  clientset,
+		keepalived: &keepalived{},
+
+		namespace:   namespace,
+		serviceName: serviceName,
+		podName:     podName,
+
+		config: make(map[string]interface{}),
+	}
+
+	tmpl, err := template.ParseFiles(keepalivedTmpl)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse keepalived config template due to: %v", err)
+	}
+	c.template = tmpl
+
+	return c, nil
+}
+
+func (c *keepalivedController) Run(period time.Duration, stopCh <-chan struct{}) {
+	go c.keepalived.Start()
+
+	go wait.Until(func() {
+		if err := c.Sync(); err != nil {
+			glog.Error(err)
+		}
+	}, period, stopCh)
+
+	<-stopCh
+}
+
+func (c *keepalivedController) Stop() error {
+	vip := c.config["vip"]
+	iface := c.config["iface"]
+
+	defer c.keepalived.Stop()
+
+	return c.removeVIP(iface.(string), vip.(string))
+}
+
+func (c *keepalivedController) Sync() error {
+	conf, err := c.fetchConfig()
+	if err != nil {
+		return err
+	}
+
+	if reflect.DeepEqual(conf, c.config) {
+		return nil
+	}
+	c.config = conf
```

Comment:
You haven't successfully changed the config yet, I'd suggest move this line to the end of the method.